README.rst
===============================
DAPIWrap: Doomworld API Wrapper
===============================

**This package is still in the beta stages, so expect updates that may break things, if you're using it.**

DAPIWrap is a fairly simple Python wrapper for the `Doomworld /idgames archive API`_. In addition to allowing you to access the API using Python, it also has a few added features, located in ``dapiwtools``, that aren't provided by the API, such as search filtering, and downloading functions, for downloading wads from the supported mirrors on the `Doomworld /idgames archive`_.

Please don't abuse this wrapper, and hammer the Doomworld servers with it.

**Note:**

It seems that the maximum number of search results the API will return is 100. Unfortunately, there is nothing I can do about this. Just a limit of the API. If you are searching for something specific, you can try being a bit more specific with your search.

Install/Uninstall with PIP_
===========================

Install
-------
::

    pip install https://github.com/Trebek/DAPIWrap/archive/master.zip

Uninstall
---------
::

    pip uninstall dapiwrap

Basic Usage
===========

These are some brief examples of how to use the functions of DAPIWrap.

Getting Wad Info
----------------

Get Wad Info Using ID
^^^^^^^^^^^^^^^^^^^^^
::

    #!/usr/bin/env python

    from dapiwrap import DAPIWrap

    daw = DAPIWrap()

    wad_info = daw.get_id(12815)

Get Wad Info Using Full Path
^^^^^^^^^^^^^^^^^^^^^^^^^^^^
::

    #!/usr/bin/env python

    from dapiwrap import DAPIWrap

    daw = DAPIWrap()

    wad_info = daw.get_file("levels/doom2/Ports/v-z/zdmcmp1.zip")

Searching
---------

Basic Search
^^^^^^^^^^^^
::

    #!/usr/bin/env python

    from dapiwrap import DAPIWrap

    daw = DAPIWrap()

    results = daw.search("zdmcmp1")

Search with Parameters & Filter
^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^
::

    #!/usr/bin/env python

    from dapiwrap import (
        DAPIWrap,
        DIRECT_DESC,
        FILTER_YEAR,
        SORT_RATING,
        TYPE_TITLE
    )

    daw = DAPIWrap()

    params = {
        "type": TYPE_TITLE,
        "sort": SORT_RATING,
        "dir": DIRECT_DESC,
        "filter": (FILTER_YEAR, 1995)
    }

    results = daw.search("test", params)

Downloading
-----------

Download a Wad, Using it's ID
^^^^^^^^^^^^^^^^^^^^^^^^^^^^^
::

    #!/usr/bin/env python

    from dapiwrap import DAPIWrap

    dl_folder = "C:\\games\\doom\\wads\\"

    daw = DAPIWrap()

    daw.download.wad_id(12815, dl_folder)

Download a Wad, Using a Full Path
^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^
::

    #!/usr/bin/env python

    from dapiwrap import DAPIWrap

    dl_folder = "C:\\games\\doom\\wads\\"

    daw = DAPIWrap()

    daw.download.file_path("levels/doom2/Ports/v-z/zdmcmp1.zip", dl_folder)

Typical Responses
=================

Getting Wad Info
----------------

Getting a specific wad's full info, using an ID or filename, typically returns a response like this (this is just an example):
::

    {
        u'age': 832402800,
        u'author': u'Some Dude',
        u'base': u'New level from scratch',
        u'bugs': u'No',
        u'buildtime': u'1 hour',
        u'credits': u'iD Software',
        u'date': u'2014-01-01',
        u'description': u'This is a brief description of the wad.',
        u'dir': u'levels/doom2/s-u/',
        u'editors': u'Doom Builder 2',
        u'email': u'SomeDude@someemail.com',
        u'filename': u'test.zip',
        u'id': 12815,
        u'idgamesurl': u'idgames://levels/doom2/s-u/test.zip',
        u'rating': 5.0,
        u'reviews': {u'review': [{u'text': u'cool map', u'vote': 5}]},
        u'size': 67005,
        u'textfile': u"The entirety of the wad's text file would be here.",
        u'title': u'Test',
        u'url': u'http://www.doomworld.com/idgames/?file=levels/doom2/s-u/test.zip',
        u'votes': 2
     }

Or ``None``, if no wad was found.

Searching
---------

A search will yield a list of more brief info for each wad found, like so (this is just an example):
::

    [
        {
            u'age': 832402800,
            u'author': u'Some Dude',
            u'date': u'2014-01-01',
            u'description': u'This is a brief description of the wad.',
            u'dir': u'levels/doom2/s-u/',
            u'email': u'SomeDude@someemail.com',
            u'filename': u'test.zip',
            u'id': 12021,
            u'idgamesurl': u'idgames://levels/doom2/s-u/test.zip',
            u'rating': 5.0,
            u'size': 67005,
            u'title': u'Test',
            u'url': u'http://www.doomworld.com/idgames/?file=levels/doom2/s-u/test.zip',
            u'votes': 2
        },
        {
            u'age': 865432674,
            u'author': u'Another Guy',
            u'date': u'2014-02-02',
            u'description': u'This is a brief description of the wad.',
            u'dir': u'levels/doom2/a-c/',
            u'email': u'SomeDude@someemail.com',
            u'filename': u'anotherwad.zip',
            u'id': 13024,
            u'idgamesurl': u'idgames://levels/doom2/a-c/anotherwad.zip',
            u'rating': 4.8,
            u'size': 76050,
            u'title': u'Another Wad',
            u'url': u'http://www.doomworld.com/idgames/?file=levels/doom2/s-u/anotherwad.zip',
            u'votes': 1
        },
    ]

Or ``[]`` (empty ``list``), if no wads were found.

Downloading
-----------

At the moment, downloading returns the closed file object.

.. _Doomworld \/idgames archive: http://www.doomworld.com/idgames/
.. _Doomworld \/idgames archive API: http://www.doomworld.com/idgames/api/
.. _PIP: https://pypi.python.org/pypi/pip/
.. _FTP return code: http://en.wikipedia.org/wiki/List_of_FTP_server_return_codes

CHANGES.rst
================
DAPIWrap Changes
================

v0.3.0 (15-06-2014)
-------------------

- Consolidated the loose functions, in ``dapiwtools`` into a couple of separate classes.
    - Added the ``IOFuncs`` class, which has the methods ``open_id_list``, ``open_json``, ``save_id_list``, and ``save_json``, for saving/opening wad id lists, and JSON data (raw, or gzipped). These functions can be accessed through a ``DAPIWrap`` instance by using ``DAPIWrap.io.<method>``, where "<method>" is the chosen method.
    - Added the ``MiscFuncs`` class, which just has some miscellaneous methods, such as ``random_wad``, ``print_wad_info``, and ``open_url``. These can be accessed through a ``DAPIWrap`` instance by using ``DAPIWrap.misc.<method>``, where "<method>" is the chosen method.
        - Changed ``random_wad``, so that now you have to provide it with an idgames archive path, or list of paths, to retrieve a random wad from. If provided a list of paths, it will pick one random path, and get a random wad from there.
- Added the method ``search_local`` to ``DAPIWrap``, for searching a given list of wad info items.
    - I added this because, apparently, the Doomworld API only returns the first 100 results when searching. However, if you know the directory you want to search in, you can retrieve a ``list`` of the entire contents of a given directory, with ``DAPIWrap.get_files``, and then search through them, by feeding the ``list`` into ``DAPIWrap.search_local``, along with a search query, and any optional parameters. It works just like ``DAPIWrap.search``, except you also feed it a list of info to search through.
    - Note though, that if you are feeding ``DAPIWrap.search_local`` a list of items retrieved by ``DAPIWrap.search`` or ``DAPIWrap.get_files``, that you can't search the credits, editors, or text files of the items, because those areas are not included in the results of those actions. They only return a brief version of each wad's info, missing those sections.
- Changed the way the ``newdir`` argument works, for the download methods in the ``Downwad`` class, in ``dapiwtools``. 
    - Was set up to create, and download files to a new directory of the same name as the file, if ``newdir=True``. Didn't have anything to do with whether or not the download path you were providing was a new directory. Not really the behaviour you would expect, I don't think. I had it set up that way for my own purposes. Now the ``newdir`` argument simply specifies whether or not you want ``DAPIWrap`` to create a new directory, if the given download path doesn't exist. If ``newdir=False``, and the download path you specify doesn't exist, Python will throw an ``IOError`` exception. By default, ``newdir=True``.
    - Downloading from an FTP server now returns the closed file object, the same way downloading from an HTTP server does.
- Fixed a misspelling in dapiwconst.
- Fixed a few errors in the documentation.

v0.2.0 (11-06-2014)
-------------------

- Got rid of some broken methods. 
    - The methods ``DAPIWrap.get_file_alt``, in ``dapiwrap``, & ``Downwad.filename``, in ``dapiwtools``, didn't really work, because multiple files, in different directories, can have the same filename. Since this method relied on searching to find the file with the given filename, it would often return multiple results, and the only way of knowing which was the right file, was to query the user through the console. I didn't like that.
- Added a few more functions to ``dapiwtools``.
    - Added ``open_id_list`` & ``save_id_list`` for saving/loading a txt file of wad IDs.
        - Can be used to compile a list of IDs, for storing/sharing, that can be opened and fed into ``DAPIWrap.get_id_list`` or ``DAPIWrap.download.id_list``
    - Added ``open_json`` & ``save_json``, for dumping/loading JSON data, either as raw JSON data, or gzipped.
        - Could be used to store retrieved wad data locally.
    - Added ``open_url`` to open the Doomworld archive page of a wad in your default browser.
- Changed the method name ``DAPIWrap.get_file``, in ``dapiwrap``, to ``DAPIWrap.get_file_path``. Was too similar to ``DAPIWrap.get_files``.
- Cleaned up ``dapiwconst``, and removed a number of useless constants, and added some useful ones.
- Removed some unused imports from ``dapiwconst``, in ``dapiwrap`` & ``dapiwtools``.

v0.1.0 (07-06-2014)
-------------------

- Initial release.

dapiwrap/dapiwrap.py
#===============================================================================
# DAPIWrap: Doomworld API Wrapper
#-------------------------------------------------------------------------------
# Version: 0.3.0
# Updated: 15-06-2014
# Author: Alex Crawford
# License: MIT
#===============================================================================

"""
DAPIWrap, short for "Doomworld API Wrapper", is a simple Python wrapper for the 
Doomworld idgames archive API.

This pacakge is still in the beta stages, so expect changes that may break
things, if you're using it.

Doomworld API documentation:
http://www.doomworld.com/idgames/api/

Doomworld idgames archive:
http://www.doomworld.com/idgames/

Note:

Please don't hammer the Doomworld API with requests. I am not sure what kind 
of traffic their API can handle, so do the right thing, and limit the number 
of requests you make. I don't want this wrapper to be used (abused) to bog down 
their server. Be cool.

This package requires the ``requests`` package:
http://docs.python-requests.org/en/latest/

"""

#===============================================================================
# Imports
#===============================================================================

import requests

from dapiwconst import (
    A_ABOUT,
    A_DBPING,
    A_GET_ID,
    A_GET_FILE,
    A_GETCONTENTS_ID,
    A_GETCONTENTS_NAME,
    A_GETDIRS_ID,
    A_GETDIRS_NAME,
    A_GETFILES_ID,
    A_GETFILES_NAME,
    A_LATESTVOTES,
    A_LATESTFILES,
    A_PARENTDIR_ID,
    A_PARENTDIR_FILE,
    A_PING,
    A_SEARCH,
    A_SEARCH_QUERY,
    A_SEARCH_DIRECT,
    A_SEARCH_SORT,
    A_SEARCH_TYPE,
    A_OUT_JSON,
    API_URL,
    DIRECT_ASC,
    DIRECT_DESC,
    SORT_DATE,
    TYPE_AUTHOR,
    TYPE_CREDITS,
    TYPE_DESCRIP,
    TYPE_EDITORS,
    TYPE_EMAIL,
    TYPE_FILE,
    TYPE_TEXT,
    TYPE_TITLE,
)

from dapiwtools import (
    Downwad,
    IOFuncs,
    MiscFuncs,
    SearchFilter
)

#===============================================================================
# DAPIWrap Class
#===============================================================================

class DAPIWrap(object):
    """The main DAPIWrap class."""

    A_NOPARAM = [
        A_ABOUT, A_DBPING, A_PING
    ]

    A_SINGLEPARAM = [
        A_GET_ID, A_GET_FILE, A_GETCONTENTS_ID, A_GETCONTENTS_NAME, 
        A_GETFILES_NAME, A_GETDIRS_ID, A_GETDIRS_NAME, A_GETFILES_ID, 
        A_LATESTFILES, A_LATESTVOTES, A_PARENTDIR_FILE
    ]

    def __init__(self, dl_folder=None):
        """
        The DAPIWrap init method.

        :param dl_folder: The location to download files to.

        """
        self.dl_folder = dl_folder

        # DAPIWrap "tools".
        self.download = Downwad(self)
        self.filter = SearchFilter()
        self.io = IOFuncs()
        self.misc = MiscFuncs(self)

    def about(self):
        """
        Get information about the Doomworld API.

        :returns: Information about the Doomworld API.

        """
        about = self.call(A_ABOUT)

        if not raw and "content" in about:
            return about["content"]
        else:
            return about

    def call(self, action, params=None):
        """
        Calls the API, using the given action/parameters.

        :param action: An action constant from ``dapiwconst``.
        :param params: Any additional parameters for the action.

        :returns: The Doomworld API response.

        """
        if action in self.A_NOPARAM:
            url = "%s%s%s" % (API_URL, action, A_OUT_JSON)
        elif action in self.A_SINGLEPARAM:
            url = "%s%s%s" % (
                API_URL, action % (
                    params
                ), A_OUT_JSON
            )
        else:
            if action == A_SEARCH:
                if type(params) == dict:
                    url = "%s%s" % (API_URL, A_SEARCH_QUERY % (params["query"]))
                    if "type" in params:
                        url += A_SEARCH_TYPE % (params["type"])
                    if "sort" in params:
                        url += A_SEARCH_SORT % (params["sort"])
                    if "dir" in params:
                        url += A_SEARCH_DIRECT % (params["dir"])
                else:
                    url = "%s%s" % (API_URL, A_SEARCH_QUERY % (params))
                url += A_OUT_JSON

        data = requests.get(url).json()

        return data

    def dbping(self):
        """
        Ping the database.

        :returns: Database status.

        """
        return self.call(A_DBPING)

    def get_contents(self, path, raw=False):
        """
        Gets the contents of a given path.

        :param path: The idgames path to get content from.
        :param raw: Whether to return the data exactly as recieved (raw), or
        to extract the contents and return just the contents in a list.

        :returns: Directory contents

        """
        contents = self.call(A_GETCONTENTS_NAME, path)

        if not raw and "content" in contents:
            return contents["content"]
        else:
            return contents

    def get_file_path(self, path, raw=False):
        """
        Gets a file from the given path.

        :param path: The full path of the file, including the filename.
        :param raw: Whether to return the data exactly as recieved (raw), or
        to extract the contents and return just the contents in a list.

        :returns: Wad info for the given file.

        """
        wad_info = self.call(A_GET_FILE, path)

        if not raw and "content" in wad_info:
            return wad_info["content"]
        else:
            return wad_info

    def get_files(self, path, raw=False):
        """
        Gets the files under a directory of the given path.

        :param path: The idgames path to return files from.
        :param raw: Whether to return the data exactly as recieved (raw), or
        to extract the contents and return just the contents in a list.

        :returns: Files under the given path.

        """
        files = self.call(A_GETFILES_NAME, path)

        if not raw and "content" in files:
            if type(files["content"]["file"]) == list:
                return files["content"]["file"]
            else:
                return [files["content"]["file"]]
        else:
            return []

    def get_dirs(self, path, raw=False):
        """
        Gets the directories under a directory of the given path.

        :param path: The idgames path to return directories from.
        :param raw: Whether to return the data exactly as recieved (raw), or
        to extract the contents and return just the contents in a list.

        :returns: Subdirectories in a given path.

        """
        dirs = self.call(A_GETDIRS_NAME, path)

        if not raw and "content" in dirs:
            return dirs["content"]
        else:
            return dirs

    def get_id(self, wad_id, raw=False):
        """
        Gets the info for a wad with the given ID.

        :param wad_id: The ID number of a wad.
        :param raw: Whether to return the data exactly as recieved (raw), or
        to extract the contents and return just the contents in a list.

        :returns: Wad info for the given ID.

        """
        wad_info = self.call(A_GET_ID, wad_id)

        if not raw and "content" in wad_info:
            return wad_info["content"]
        else:
            return wad_info

    def get_id_list(self, id_list, raw=False):
        """
        Gets the info for a wad with the given ID.

        :param id_list: A list of wad ID numbers.
        :param raw: Whether to return the data exactly as recieved (raw), or
        to extract the contents and return just the contents in a list.

        :returns: Wad info for the given list of IDs.

        """
        results = []

        for wad_id in id_list:
            wad_info = self.call(A_GET_ID, wad_id)
            if not raw and "content" in wad_info:
                results.append(wad_info["content"])
            else:
                return wad_info

        return results

    def get_latestfiles(self, limit=10):
        """
        Get the latest uploaded files.

        :param limit: The number of items to retrieve.

        :returns: The latest files.

        """
        latest = self.call(A_LATESTFILES, limit)

        if not raw and "content" in latest:
            return latest["content"]
        else:
            return latest

    def get_latestvotes(self, limit=10):
        """
        Get the latest votes.

        :param limit: The number of items to retrieve.

        :returns: The latest votes.

        """
        latest = self.call(A_LATESTVOTES, limit)

        if not raw and "content" in latest:
            return latest["content"]
        else:
            return latest

    def get_parent_dir(self, path, raw=False):
        """
        Gets the parent directory for a given path.

        :param path: The idgames path to return the parent from.
        :param raw: Whether to return the data exactly as recieved (raw), or
        to extract the contents and return just the contents in a list.

        :returns: The parent directory of the given directory path.

        """
        parent_dir = self.call(A_PARENTDIR_FILE, path)

        if not raw and "content" in parent_dir:
            return parent_dir["content"]
        else:
            return parent_dir

    def ping(self):
        """
        Ping the Doomworld server.

        :returns: The Doomworld server status.

        """
        return self.call(A_PING)

    def search(self, query, params={}):
        """
        Search the idgames archive.

        :param query: The search query.
        :param params: A ``dict`` of search parameters and/or filters.

        ``params`` can be one, or one of each of these keys/values:

        :param type: The type of search (what field your searching in).
            |Accepted Values:
            |TYPE_AUTHOR, TYPE_CREDITS, TYPE_DESCRIP, TYPE_EDITORS,
            |TYPE_EMAIL, TYPE_FILE, TYPE_TEXT, TYPE_TITLE            
        :param sort: What to sort the results by.
            |Accepted Values:
            |SORT_DATE, SORT_FILE, SORT_RATING, SORT_SIZE
        :param dir: The direction to order the results.
            |Accepted Values:
            |DIRECT_ASC, DIRECT_DESC
        :param filter: A filter, or filters (in a tuple/list) to apply to 
            the results. If you are using filters, you must also provide values 
            for the filters, and package them both into a tuple, or list. For 
            example, if you wanted to filter by year - 1994 in this case - you 
            would use this as a value for filter: 
            |(FILTER_YEAR, 1994)
            If you wanted to use more than one filter, in this case by year 
            and game, you would use this:
            |((FILTER_YEAR, 1994), (FILTER_GAME, DOOM))
            Accepted Values:
            |FILTER_GAME, FILTER_GYR, FILTER_YEAR, FILTER_RATING

        :returns: A list of search results.

        """
        params["query"] = query

        results = self.call(A_SEARCH, params)

        if "raw" not in params and "content" in results:
            if type(results["content"]["file"]) == list:
                results = results["content"]["file"]
            else:
                results = [results["content"]["file"]]
            if "filter" in params:
                if type(params["filter"][0]) not in (tuple, list):
                    srchfilter = params["filter"][0]
                    filterargs = params["filter"][1]
                    results = self.filter.filters[srchfilter](
                        results, filterargs
                    )
                else:
                    for item in params["filter"]:
                        srchfilter = item[0]
                        filterargs = item[1]
                        results = self.filter.filters[srchfilter](
                            results, filterargs
                        )
            return results
        return []

    def search_author(self, query, params={}):
        """
        Search the idgames archive for the given author, using the 
        default search parameters.

        :param query: The search query.
        :param params: Optional. A ``dict`` of parameters and/or filters.

        :returns: The search results.

        """
        if not params:
            return self.search(query, {"type": TYPE_AUTHOR})
        else:
            params["type"] = TYPE_AUTHOR
            return self.search(query, params)

    def search_credits(self, query, params={}):
        """
        Search the credits of the files in the idgames archive, using the 
        default search parameters.

        :param query: The search query.
        :param params: Optional. A ``dict`` of parameters and/or filters.

        :returns: The search results.

        """
        if not params:
            return self.search(query, {"type": TYPE_CREDITS})
        else:
            params["type"] = TYPE_CREDITS
            return self.search(query, params)

    def search_descrip(self, query, params={}):
        """
        Search the descriptions of files in the idgames archive, using 
        the default search parameters.

        :param query: The search query.
        :param params: Optional. A ``dict`` of parameters and/or filters.

        :returns: The search results.

        """
        if not params:
            return self.search(query, {"type": TYPE_DESCRIP})
        else:
            params["type"] = TYPE_DESCRIP
            return self.search(query, params)

    def search_editors(self, query, params={}):
        """
        Search the idgames archive for the given editors, using the 
        default search parameters.

        :param query: The search query.
        :param params: Optional. A ``dict`` of parameters and/or filters.

        :returns: The search results.

        """
        if not params:
            return self.search(query, {"type": TYPE_EDITORS})
        else:
            params["type"] = TYPE_EDITORS
            return self.search(query, params)

    def search_email(self, query, params={}):
        """
        Search the idgames archive for the given email, using the 
        default search parameters.

        :param query: The search query.
        :param params: Optional. A ``dict`` of parameters and/or filters.

        :returns: The search results.

        """
        if not params:
            return self.search(query, {"type": TYPE_EMAIL})
        else:
            params["type"] = TYPE_EMAIL
            return self.search(query, params)

    def search_local(self, items, query, params={}):
        """        
        Search the given items. Can't search the credits, editors, or textfiles 
        of the given items, if they were retrieved using ``DAPIWrap.search``, 
        or ``DAPIWrap.get_files``, because these actions don't return the full 
        info for each wad. Only a brief version, missing those areas mentioned.
        However, if the items were retrieved using ``DAPIWrap.get_id`` or
        ``DAPIWrap.get_file_path``, then you can search those areas as well.

        :param items: A list of wad info to search through.
        :param query: The search query.
        :param params: A ``dict`` of search parameters and/or filters.

        :returns: A list of search results.

        """
        query_lower = query.lower()

        srchtype = params.get("type")
        srchsort = params.get("sort")
        srchdir = params.get("dir")
        srchfilters = params.get("filter")

        # Set up default search parameters, if none given.
        if not srchtype:
            srchtype = TYPE_FILE
        if not srchsort:
            srchsort = SORT_DATE
        if not srchdir:
            srchdir = DIRECT_ASC

        # Search through given items, build a list of results.
        try:
            local_results = [
                x for x in items if query_lower in x[srchtype].lower()
            ]
        except AttributeError:
            pass

        # Filter the results, if called for.
        if srchfilters:
            local_results = self.filter.chain(local_results, srchfilters)

        # Sort the results.
        local_results = sorted(local_results, key=lambda x: x[srchsort])

        # Reverse the results order, if called for.
        if srchdir == DIRECT_DESC:
            return local_results[::-1]
        else:
            return local_results

    def search_text(self, query, params={}):
        """
        Search the idgames archive for the given title, using the 
        default search parameters.

        :param query: The search query.
        :param params: Optional. A ``dict`` of parameters and/or filters.

        :returns: The search results.

        """
        if not params:
            return self.search(query, {"type": TYPE_TEXT})
        else:
            params["type"] = TYPE_TEXT
            return self.search(query, params)

    def search_title(self, query, params={}):
        """
        Search the idgames archive for the given title, using the 
        default search parameters.

        :param query: The search query.
        :param params: Optional. A ``dict`` of parameters and/or filters.

        :returns: The search results.

        """
        if not params:
            return self.search(query, {"type": TYPE_TITLE})
        else:
            params["type"] = TYPE_TITLE
            return self.search(query, params)

#===============================================================================
# If Main
#===============================================================================

if __name__ == '__main__':
    print "You're doing it wrong."

dapiwrap/dapiwconst.py
#===============================================================================
# DAPIWConst Global Constants
#-------------------------------------------------------------------------------
# Version: 0.3.0
# Updated: 15-06-2014
# Author: Alex Crawford
# License: MIT
#===============================================================================

"""
This file contains all of the constants that are used by DAPIWrap and it's
associated tools, in ``dapiwtools``. I just figured it would be easier to keep 
them all consolidated. Users can just import the ones they need.

"""

#-------------------------------------------------------------------------------
# API URL and API Actions
#-------------------------------------------------------------------------------

# Doomworld API url
API_URL = "http://www.doomworld.com/idgames/api/api.php?action="

# API Actions
A_ABOUT = "about"
A_DBPING = "dbping"
A_GET_ID = "get&id=%d"
A_GET_FILE = "get&file=%s"
A_GETCONTENTS_ID = "getcontents&id=%d"
A_GETCONTENTS_NAME = "getcontents&name=%s"
A_GETDIRS_ID = "getdirs&id=%d"
A_GETDIRS_NAME = "getdirs&name=%s"
A_GETFILES_ID = "getfiles&id=%d"
A_GETFILES_NAME = "getfiles&name=%s"
A_LATESTVOTES = "latestvotes&limit=%d"
A_LATESTFILES = "latestfiles&limit=%d"
A_OUT_JSON = "&out=json"
A_PARENTDIR_ID = "getparentdir&id=%d"
A_PARENTDIR_FILE = "getparentdir&name=%s"
A_PING = "ping"
A_SEARCH = "search&query=%s&type=%s&sort=%s&dir=%s"
A_SEARCH_QUERY = "search&query=%s"
A_SEARCH_DIRECT = "&dir=%s"
A_SEARCH_SORT = "&sort=%s"
A_SEARCH_TYPE = "&type=%s"

#-------------------------------------------------------------------------------
# Download Servers
#-------------------------------------------------------------------------------

# FTP Download servers
DL_FTP_GERMANY = "ftp.fu-berlin.de"
DL_FTP_GREECE = "ftp.ntua.gr"
DL_FTP_TEXAS = "ftp.mancubus.net"
DL_FTP = [DL_FTP_GERMANY, DL_FTP_GREECE, DL_FTP_TEXAS]

# HTTP Download servers
DL_FLORIDA = "http://www.gamers.org/pub/"
DL_GREECE = "http://ftp.ntua.gr/pub/vendors/"
DL_NEWYORK = "http://youfailit.net/pub/"
DL_TEXAS = "http://ftp.mancubus.net/pub/"
DL_HTTP = [DL_FLORIDA, DL_GREECE, DL_NEWYORK, DL_TEXAS]

#-------------------------------------------------------------------------------
# FTP Specific Paths
#-------------------------------------------------------------------------------

FTP_DIR_GERMANY = "pc/games/idgames/"
FTP_DIR_GREECE = "pub/vendors/idgames/"
FTP_DIR_TEXAS = "pub/idgames/"

#-------------------------------------------------------------------------------
# /idgames Paths and Game Names
#-------------------------------------------------------------------------------

# idgames root
IDGAMES = "idgames/"

# Game path names
DOOM = "doom"
DOOM2 = "doom2"
HACX = "hacx"
HERETIC = "heretic"
HEXEN = "hexen"
STRIFE = "strife"
GAMES = [DOOM, DOOM2, HACX, HERETIC, HEXEN, STRIFE]

# Level paths
LVLS = "levels/%s/"
LVLS_09 = "levels/%s/0-9/"
LVLS_AC = "levels/%s/a-c/"
LVLS_DF = "levels/%s/d-f/"
LVLS_GI = "levels/%s/g-i/"
LVLS_JL = "levels/%s/j-l/"
LVLS_MO = "levels/%s/m-o/"
LVLS_PR = "levels/%s/p-r/"
LVLS_SU = "levels/%s/s-u/"
LVLS_VZ = "levels/%s/v-z/"
LVLS_DEATH = "levels/%s/deathmatch/"
LVLS_DEATH_09 = "levels/%s/deathmatch/0-9/"
LVLS_DEATH_AC = "levels/%s/deathmatch/a-c/"
LVLS_DEATH_DF = "levels/%s/deathmatch/d-f/"
LVLS_DEATH_GI = "levels/%s/deathmatch/g-i/"
LVLS_DEATH_JL = "levels/%s/deathmatch/j-l/"
LVLS_DEATH_MO = "levels/%s/deathmatch/m-o/"
LVLS_DEATH_PR = "levels/%s/deathmatch/p-r/"
LVLS_DEATH_SU = "levels/%s/deathmatch/s-u/"
LVLS_DEATH_VZ = "levels/%s/deathmatch/v-z/"
LVLS_DEATH_MEGA = "levels/%s/deathmatch/megawads/"
LVLS_DEATH_PORTS_09 = "levels/%s/deathmatch/Ports/0-9/"
LVLS_DEATH_PORTS_AC = "levels/%s/deathmatch/Ports/a-c/"
LVLS_DEATH_PORTS_DF = "levels/%s/deathmatch/Ports/d-f/"
LVLS_DEATH_PORTS_GI = "levels/%s/deathmatch/Ports/g-i/"
LVLS_DEATH_PORTS_JL = "levels/%s/deathmatch/Ports/j-l/"
LVLS_DEATH_PORTS_MO = "levels/%s/deathmatch/Ports/m-o/"
LVLS_DEATH_PORTS_PR = "levels/%s/deathmatch/Ports/p-r/"
LVLS_DEATH_PORTS_SU = "levels/%s/deathmatch/Ports/s-u/"
LVLS_DEATH_PORTS_VZ = "levels/%s/deathmatch/Ports/v-z/"
LVLS_DEATH_PORTS_MEGA = "levels/%s/deathmatch/Ports/megawads/"
LVLS_MEGA = "levels/%s/megawads/"
LVLS_PORTS = "levels/%s/Ports/"
LVLS_PORTS_09 = "levels/%s/Ports/0-9/"
LVLS_PORTS_AC = "levels/%s/Ports/a-c/"
LVLS_PORTS_DF = "levels/%s/Ports/d-f/"
LVLS_PORTS_GI = "levels/%s/Ports/g-i/"
LVLS_PORTS_JL = "levels/%s/Ports/j-l/"
LVLS_PORTS_MO = "levels/%s/Ports/m-o/"
LVLS_PORTS_PR = "levels/%s/Ports/p-r/"
LVLS_PORTS_SU = "levels/%s/Ports/s-u/"
LVLS_PORTS_VZ = "levels/%s/Ports/v-z/"
LVLS_PORTS_MEGA = "levels/%s/Ports/megawads/"

#-------------------------------------------------------------------------------
# Level Path Lists
#-------------------------------------------------------------------------------

# Alphabetical paths under "levels/<GAME>/"
LVLS_ALPHA = [
    LVLS_09, LVLS_AC, LVLS_DF, LVLS_GI, LVLS_JL, LVLS_MO, LVLS_PR, LVLS_SU,
    LVLS_VZ
]

# Alphabetical paths under "levels/<GAME>/deathmatch/"
LVLS_DEATH_ALPHA = [
    LVLS_DEATH_09, LVLS_DEATH_AC, LVLS_DEATH_DF, LVLS_DEATH_GI, LVLS_DEATH_JL,
    LVLS_DEATH_MO, LVLS_DEATH_PR, LVLS_DEATH_SU, LVLS_DEATH_VZ
]

# Alphabetical paths under "levels/<GAME>/deathmatch/Ports/"
LVLS_DEATH_PORTS_ALPHA = [
    LVLS_DEATH_PORTS_09, LVLS_DEATH_PORTS_AC, LVLS_DEATH_PORTS_DF,
    LVLS_DEATH_PORTS_GI, LVLS_DEATH_PORTS_JL, LVLS_DEATH_PORTS_MO,
    LVLS_DEATH_PORTS_PR, LVLS_DEATH_PORTS_SU, LVLS_DEATH_PORTS_VZ
]

# Alphabetical paths under "levels/<GAME>/Ports/"
LVLS_PORTS_ALPHA = [
    LVLS_PORTS_09, LVLS_PORTS_AC, LVLS_PORTS_DF, LVLS_PORTS_GI, LVLS_PORTS_JL, 
    LVLS_PORTS_MO, LVLS_PORTS_PR, LVLS_PORTS_SU, LVLS_PORTS_VZ
]

# All paths under "levels/"
LVLS_ALL = (
    LVLS_ALPHA + [LVLS_MEGA] + LVLS_DEATH_ALPHA + [LVLS_DEATH_MEGA] + 
    LVLS_DEATH_PORTS_ALPHA + [LVLS_DEATH_PORTS_MEGA] + LVLS_PORTS_ALPHA + 
    [LVLS_PORTS_MEGA]
)

#-------------------------------------------------------------------------------
# Search Parameters
#-------------------------------------------------------------------------------

# Search types
TYPE_AUTHOR = "author"
TYPE_CREDITS = "credits"
TYPE_DESCRIP = "description"
TYPE_EDITORS = "editors"
TYPE_EMAIL = "email"
TYPE_FILE = "filename"
TYPE_TEXT = "textfile"
TYPE_TITLE = "title"

# Search sorting methods
SORT_DATE = "date"
SORT_FILE = "filename"
SORT_RATING = "rating"
SORT_SIZE = "size"

# Sort direction
DIRECT_ASC = "asc"
DIRECT_DESC = "desc"

#-------------------------------------------------------------------------------
# Filter Stuff
#-------------------------------------------------------------------------------

# Filter names
FILTER_DATE = "filter:date"
FILTER_GAME = "filter:game"
FILTER_GYR = "filter:gyr"
FILTER_RATING = "filter:rating"
FILTER_SIZE = "filter:size"
FILTER_VOTES = "filter:votes"
FILTER_YEAR = "filter:year"
FILTERS = [
    FILTER_DATE, FILTER_GAME, FILTER_RATING, 
    FILTER_SIZE, FILTER_VOTES, FILTER_YEAR
]

#===============================================================================
# If Main
#===============================================================================

if __name__ == '__main__':
    print "You're doing it wrong."